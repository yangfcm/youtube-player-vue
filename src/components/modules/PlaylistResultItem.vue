<template>
  <div class="ui items">
    <div class="item">
      <router-link
        class="ui small image"
        style="position: relative"
        :to="`/plalist/${playlist.id.playlistId}`"
      >
        <img :src="playlist.snippet.thumbnails.medium.url" />
        <div class="app-image-cover">
          <p>
            <i class="list icon"></i>
            Play List
          </p>
        </div>
      </router-link>
      <div class="content">
        <router-link
          class="header"
          :to="`/playlist/${playlist.id.playlistId}`"
        >{{ playlist.snippet.title }}</router-link>
        <div class="meta">
          <router-link
            :to="`/channel/${playlist.snippet.channelId}`"
          >{{ playlist.snippet.channelTitle }}</router-link>
          <div class="ui right floated">
            <i class="calendar plus icon"></i>
            {{ playlist.snippet.publishedAt | date }}
          </div>
        </div>
        <div class="description" style="overflow: auto;">{{ playlist.snippet.description }}</div>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  props: ["playlist"],
};
</script>
<style scoped>
.app-image-cover {
  height: 100%;
  width: 100%;
  color: white;
  position: absolute;
  z-index: 10;
  background: rgba(0, 0, 0, 0.4);
  top: 0px;
  display: flex;
  justify-content: center;
  align-items: center;
}
</style>