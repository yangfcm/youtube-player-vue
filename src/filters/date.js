/**
 * Convert a date object to human-readable string
 */
export default (date) => {
  const dateObj = new Date(date);
  const day = dateObj.getDate();
  const month = new Intl.DateTimeFormat("en-US", {
    month: "short",
  }).format(dateObj);
  const year = dateObj.getFullYear();
  return `${day} ${month}, ${year}`;
};
