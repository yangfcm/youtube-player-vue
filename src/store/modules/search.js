import axios from "../../settings";
import { SEARCH, CATCH_SEARCH_ERROR } from "../types";
import parseError from "../helpers/parseError";

const state = {
  searchResults: null,
  searchError: null,
};

const getters = {
  searchErrorMessage: (state) => {
    return parseError(state.searchError, "Failed to proceed searching");
  },
};

const mutations = {
  SEARCH: (state, payload) => {
    state.searchResults = payload;
    state.searchError = null;
  },
  CATCH_SEARCH_ERROR: (state, payload) => {
    state.searchError = payload;
  },
};

const actions = {
  search: async (context, keyword, pageToken) => {
    try {
      const response = await axios.get("/search", {
        params: {
          ...axios.defaults.params,
          q: keyword,
          part: "snippet",
          maxResults: 15,
          pageToken,
        },
      });
      context.commit(SEARCH, response.data);
    } catch (err) {
      context.commit(CATCH_SEARCH_ERROR, err);
    }
  },
};

export default {
  state,
  getters,
  mutations,
  actions,
};
