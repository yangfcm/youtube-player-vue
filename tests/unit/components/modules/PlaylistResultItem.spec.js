import { shallowMount, RouterLinkStub, createLocalVue } from "@vue/test-utils";
import PlaylistResultItem from "@/components/modules/PlaylistResultItem";
import { searchResultResponse } from "../../store/fixtures/search";
import date from "@/filters/date";

describe("Test PlaylistResultItem component", () => {
  let wrapper;
  const localVue = createLocalVue();
  localVue.filter("date", date);

  const playlist = searchResultResponse.items.find(
    (item) => item.id.kind === "youtube#playlist"
  );

  beforeEach(() => {
    wrapper = shallowMount(PlaylistResultItem, {
      propsData: {
        playlist,
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
      localVue,
    });
  });
  afterEach(() => {
    wrapper.destroy();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correct content", () => {
    expect(wrapper.find("a.image").props().to).toBe(
      `/plalist/${playlist.id.playlistId}`
    );
    expect(wrapper.find("img").attributes("src")).toBe(
      playlist.snippet.thumbnails.medium.url
    );
    expect(wrapper.find(".header").text()).toContain(playlist.snippet.title);
    expect(wrapper.find(".description").text()).toContain(
      playlist.snippet.description
    );
    expect(wrapper.find("div.meta a").props().to).toBe(
      `/channel/${playlist.snippet.channelId}`
    );
  });
});
