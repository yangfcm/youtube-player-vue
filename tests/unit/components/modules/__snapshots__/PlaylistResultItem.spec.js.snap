// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PlaylistResultItem component should match snapshot 1`] = `
<div class="ui items">
  <div class="item"><a class="ui small image" style="position: relative;"><img src="https://i.ytimg.com/vi/IDZyZMYn3H0/mqdefault.jpg">
      <div class="app-image-cover">
        <p><i class="list icon"></i>
          Play List
        </p>
      </div>
    </a>
    <div class="content"><a class="header">How to Become a Google Developer Expert (GDE)</a>
      <div class="meta"><a>Fireship</a>
        <div class="ui right floated"><i class="calendar plus icon"></i>
          24 May, 2019
        </div>
      </div>
      <div class="description" style="overflow: auto;">What does it take to become a Google Developer Expert? Learn about my own experience becoming a GDE with practical advice for working towards this goal.</div>
    </div>
  </div>
</div>
`;
